File: src/main/java/pl/project13/tinytermpm/marshalling/JAXBUtil.java
package pl.project13.tinytermpm.marshalling;

import javax.xml.bind.JAXBContext;
import javax.xml.bind.JAXBException;
import javax.xml.bind.Marshaller;
import javax.xml.bind.Unmarshaller;
import java.io.ByteArrayInputStream;
import java.io.StringWriter;

public class JAXBUtil {

    JAXBUtil() {
    }

    static public <T> T unmarshal(Class<T> clazz, String response) throws JAXBException {
        Unmarshaller unmarshaller = createUnmarshaller(clazz);

        T unmarshal = unmarshal(clazz, response, unmarshaller);
        return unmarshal;
    }

    static public <T> String marshal(T objectToMarshal) {
        try {
            JAXBContext ctx = JAXBContext.newInstance(objectToMarshal.getClass());
            Marshaller marshaller = ctx.createMarshaller();
            StringWriter writer = new StringWriter(500);

            marshaller.marshal(objectToMarshal, writer);

            String marshaled = writer.toString();
            return marshaled;
        } catch (JAXBException e) {
            throw new RuntimeException(e);
        }
    }

    static private <T> T unmarshal(Class<T> clazz, String response, Unmarshaller unmarshaller) throws JAXBException {
        ByteArrayInputStream is = new ByteArrayInputStream(response.getBytes());
        Object unmarshalled = unmarshaller.unmarshal(is);
        return clazz.cast(unmarshalled);
    }

    static private <T> Unmarshaller createUnmarshaller(Class<T> clazz) throws JAXBException {
        JAXBContext ctx = JAXBContext.newInstance(clazz);
        return ctx.createUnmarshaller();
    }
}

File: src/main/scala/pl/project13/tinytermpm/marshalling/JAXBUtil.scala
package pl.project13.tinytermpm.marshalling

import javax.xml.bind.JAXBContext
import javax.xml.bind.JAXBException
import javax.xml.bind.Marshaller
import javax.xml.bind.Unmarshaller
import java.io.ByteArrayInputStream
import java.io.StringWriter

object JAXBUtil {
  def unmarshal[T](response: String)(implicit manifest: Manifest[T]): T = {
    val unmarshaller = createUnmarshaller(manifest.erasure)

    unmarshal(response, unmarshaller)
  }

  def marshal[T](objectToMarshal: T)(implicit manifest: Manifest[T]): String = {
    try {
      val ctx = JAXBContext.newInstance(manifest.erasure)
      val marshaller = ctx.createMarshaller
      val writer = new StringWriter(500)

      marshaller.marshal(objectToMarshal, writer)
      writer.toString
    } catch {
      case e: JAXBException => throw new RuntimeException(e)
    }
  }

  private def unmarshal[T](response: String, unmarshaller: Unmarshaller)(implicit manifest: Manifest[T]): T = {
    val is = new ByteArrayInputStream(response.getBytes)

    unmarshaller.unmarshal(is).asInstanceOf[T]
  }

  private def createUnmarshaller[T](clazz: Class[T]): Unmarshaller = {
    val ctx = JAXBContext.newInstance(clazz)
    ctx.createUnmarshaller
  }
}